Common/Securities/DynamicSecurityData.cs

```diff
@@ -0,0 +1,182 @@
+            // create a List<TradeBar> so that when accessed via GetAll<T> or .TradeBar
+            // the types line up as expected
+            var list = (IList) Activator.CreateInstance(typeof(List<>).MakeGenericType(dataType));
+            foreach (var datum in data)
+            {
+                // if the element type and dataType aren't in alignment we'll get an invalid cast exception here
+                list.Add(datum);
+            }
+
+            SetProperty(dataType.Name, list);
+        }
+
+        /// <summary>
+        /// Gets the last item in the data list for the specified type
+        /// </summary>
+        public T Get<T>()
+        {
+            var list = GetAll<T>();
+            return list.LastOrDefault();
+        }
+
+        /// <summary>
+        /// Gets the data list for the specified type
+        /// </summary>
+        public IReadOnlyList<T> GetAll<T>()
+        {
+            var data = GetProperty(typeof(T).Name);
+
+            var list = data as IReadOnlyList<T>;
+            if (list != null)
+            {
+                return list;
+            }
+
+            var baseDataList = data as IReadOnlyList<BaseData>;
+            if (baseDataList != null)
+            {
+                list = new List<T>(baseDataList.OfType<T>());
+                StoreData(typeof(T), list);
+                return list;
+            }
+
+            throw new InvalidOperationException(
+                $"Expected a list with type '{typeof(IReadOnlyList<T>).GetBetterTypeName()}' " +
+                $"but found type '{data.GetType().GetBetterTypeName()}"
+            );
+        }
+
+        /// <summary>
+        /// Sets the property with the specified name to the value. This is a case-insensitve search.
+        /// </summary>
+        /// <param name="name">The property name to set</param>
+        /// <param name="value">The new property value</param>
+        /// <returns>Returns the input value back to the caller</returns>
+        public object SetProperty(string name, object value)
+        {
+            _storage[name] = value;
+            return value;
+        }
+
+        /// <summary>
+        /// Gets the property's value with the specified name. This is a case-insensitve search.
+        /// </summary>
+        /// <param name="name">The property name to access</param>
+        /// <returns>object value of BaseData</returns>
+        public object GetProperty(string name)
+        {
+            object value;
+            if (_storage.TryGetValue(name, out value))
+            {
+                return value;
+            }
+
+            // check to see if the requested name matches one of the algorithm registered data types and if
+            // so, we'll return a new empty list. this precludes us from always needing to check HasData<T>
+            foreach (var type in _registeredTypes.GetRegisteredDataTypes())
+            {
+                if (type.Name == name)
```

Comment:
Here's where we enforce default value semantics. Consider the first time step and assume we haven't received an `TradeBar` yet for a particular security, but we've requested _at least_ one `TradeBar` subscription already. If the algo does `security.Data.TradeBar` it will catch at this condition and we'll use activator to create a new empty `List<TradeBar>` even though there's technically nothing here. We figured this default value behavior was better than throwing `KeyNotFoundException` -- thanks @jaredbroad for thinking of this case 👍